### PR TITLE
Refactor Dynamo (custom op) integration code

### PR DESCRIPTION
**Pull request by yeounoh**
Refactor after #5712 cc @wonjoolee95 @JackCaoG

### CODE REVIEW

**Comment by yeounoh** [comment]
@JackCaoG I see this test failing, but other dynamo tests are still passing
```
======================================================================
FAIL: test_dynamo_input_sharding_threashold (__main__.DynamoSpmdInferenceTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test/spmd/test_dynamo_spmd.py", line 174, in test_dynamo_input_sharding_threashold
    self.assertTrue(torch_xla._XLAC._is_placecholder(dynamo_res))
AssertionError: False is not true

----------------------------------------------------------------------
```
not sure what's happening, this fails consistently with/without the dynamo custom op. cc @wonjoolee95

**Comment by JackCaoG** [comment]
Let me take a look later today

**Comment by yeounoh** [comment]
> Let me take a look later today

The test requires multiple devices & non-GPU. So it won't run on the CI -- needs to test locally.

**Comment by JackCaoG** [comment]
OK.. I can tell you what's support to happen. `XLA_DYNAMO_INPUT_SHARDING_CHECK_THRESHOLD` controls how many times dynamo will check if input sharding changes. https://github.com/pytorch/xla/blob/6e661304e270c6a1a864700e04d5d44235c43f3b/torch_xla/core/dynamo_bridge.py#L364-L378

After THRESHOLD is reached, the dynamo won't check about input sharding. If we changed the input sharding, we will try to execute a compiled program with input that has different sharding, so it will crash. The check in the test was to make sure crash actually happened. I think you can just step through the test, you should see some C++ exception long during `try catch`.

File: torch_xla/distributed/spmd/xla_sharding.py [hunk 1]
```diff
@@ -508,29 +505,19 @@ def mark_sharding(t: Union[torch.Tensor, XLAShardedTensor],
   assert len(t.shape) == len(partition_spec), \
     f"Partition spec length ({len(partition_spec)}) should be equal to the input rank ({len(t.shape)})."
 
+  op_sharding = mesh.get_op_sharding(partition_spec)
+  tile_assignment, group_assignment, replication_groups, sharding_type = _extract_op_sharding_specs(
```
Comment (wonjoolee95) [inline torch_xla/distributed/spmd/xla_sharding.py:509]: These op sharding specs are only needed for dynamo custom op, so maybe we can move this extraction under the if statement below?

File: torch_xla/csrc/xla_sharding_util.h [hunk 2]
```diff
@@ -151,15 +151,16 @@ class ShardingUtil {
       const std::vector<std::string>& devices,
       const XLATensor::ShardingSpecPtr& sharding_spec);
 
-  static void xla_mark_sharding(const at::Tensor& input,
-                                xla::OpSharding sharding);
+  //////////////////////////// Dynamo Integration ////////////////////////////
+
+  static void XlaMarkSharding(const at::Tensor& input,
```
Comment (wonjoolee95) [inline torch_xla/csrc/xla_sharding_util.h:156]: nit: the `XlaMarkSharding` function itself here is not actually part of Dynamo Integration as the comment suggests, right? Only the `XlaMarkShardingDynamoCustomOp` is part of Dynamo Integration.

File: torch_xla/distributed/spmd/xla_sharding.py (hunk 1 above)
Comment (yeounoh) [reply torch_xla/distributed/spmd/xla_sharding.py:509]: Good catch, done

File: torch_xla/csrc/xla_sharding_util.h (hunk 2 above)
Comment (yeounoh) [reply torch_xla/csrc/xla_sharding_util.h:156]: SGTM

**Comment by yeounoh** [comment]
> OK.. I can tell you what's support to happen. `XLA_DYNAMO_INPUT_SHARDING_CHECK_THRESHOLD` controls how many times dynamo will check if input sharding changes.
> 
> https://github.com/pytorch/xla/blob/6e661304e270c6a1a864700e04d5d44235c43f3b/torch_xla/core/dynamo_bridge.py#L364-L378
> 
> After THRESHOLD is reached, the dynamo won't check about input sharding. If we changed the input sharding, we will try to execute a compiled program with input that has different sharding, so it will crash. The check in the test was to make sure crash actually happened. I think you can just step through the test, you should see some C++ exception long during `try catch`.

Synced offline with @JackCaoG he would help follow up. TLDR, it's recompiling, and we need to check if the threashold is being enforced.
```
# TODO(yeounoh) - this actually returns False, which means that the program was recompiled
    # with the new sharding change. We expect it to be True after a crash without
    # recompilation. Disabling the test until we debug.
    #self.assertTrue(torch_xla._XLAC._is_placecholder(dynamo_res))
    ```

**Comment by yeounoh** [comment]
@JackCaoG @wonjoolee95 we can review & land this refacotring PR, I won't address the test issue here (also, the test is not being run in the CPU/GPU CI).

**Comment by yeounoh** [comment]
Ok, found another test regression -- `test_mark_sharding_inside_compile` works with `torch` nightly from `11/14/2023` but started failng with the latest (`11/22/2023`):
```
======================================================================
FAIL: test_mark_sharding_inside_compile (__main__.DynamoSpmdInferenceTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test/spmd/test_dynamo_spmd.py", line 232, in test_mark_sharding_inside_compile
    dynamo_res = dynamo_linear(xla_x)
  File "/usr/local/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/torch/_dynamo/eval_frame.py", line 489, in _fn
    return fn(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
  File "test/spmd/test_dynamo_spmd.py", line 32, in forward
    xs.mark_sharding(
  File "/usr/local/lib/python3.8/site-packages/torch_xla-2.2.0+git4e506b7-py3.8-linux-x86_64.egg/torch_xla/runtime.py", line 78, in wrapper
    if not using_pjrt():
  File "/usr/local/lib/python3.8/site-packages/torch_xla-2.2.0+git4e506b7-py3.8-linux-x86_64.egg/torch_xla/runtime.py", line 82, in resume_in_wrapper
    return fn(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/torch_xla-2.2.0+git4e506b7-py3.8-linux-x86_64.egg/torch_xla/distributed/spmd/xla_sharding.py", line 499, in mark_sharding
    num_devices = xr.global_runtime_device_count()
  File "/usr/local/lib/python3.8/site-packages/torch_xla-2.2.0+git4e506b7-py3.8-linux-x86_64.egg/torch_xla/distributed/spmd/xla_sharding.py", line 499, in resume_in_mark_sharding
    num_devices = xr.global_runtime_device_count()
  File "/usr/local/lib/python3.8/site-packages/torch_xla-2.2.0+git4e506b7-py3.8-linux-x86_64.egg/torch_xla/distributed/spmd/xla_sharding.py", line 510, in resume_in_mark_sharding
    tile_assignment, group_assignment, replication_groups, sharding_type = _extract_op_sharding_specs(
  File "/usr/local/lib/python3.8/site-packages/torch_xla-2.2.0+git4e506b7-py3.8-linux-x86_64.egg/torch_xla/distributed/spmd/xla_sharding.py", line 413, in _extract_op_sharding_specs
    def _extract_op_sharding_specs(mesh: Mesh, partition_spec: Tuple):
  File "/usr/local/lib/python3.8/site-packages/torch/_dynamo/eval_frame.py", line 489, in _fn
    return fn(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/torch/_dynamo/external_utils.py", line 17, in inner
    return fn(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/torch/_functorch/aot_autograd.py", line 4960, in forward
    return compiled_fn(full_args)
  File "/usr/local/lib/python3.8/site-packages/torch/_functorch/aot_autograd.py", line 2017, in g
    return f(*args)
  File "/usr/local/lib/python3.8/site-packages/torch/_functorch/aot_autograd.py", line 3164, in runtime_wrapper
    all_outs = call_func_with_args(
  File "/usr/local/lib/python3.8/site-packages/torch/_functorch/aot_autograd.py", line 2041, in call_func_with_args
    out = normalize_as_list(f(args))
  File "/usr/local/lib/python3.8/site-packages/torch/_functorch/aot_autograd.py", line 2145, in rng_functionalization_wrapper
    return compiled_fw(args)
  File "/usr/local/lib/python3.8/site-packages/torch/_functorch/aot_autograd.py", line 2017, in g
    return f(*args)
  File "/usr/local/lib/python3.8/site-packages/torch/_dynamo/backends/torchxla.py", line 49, in fwd
    compiled_graph = bridge.extract_compiled_graph(model, args)
  File "/usr/local/lib/python3.8/site-packages/torch_xla-2.2.0+git4e506b7-py3.8-linux-x86_64.egg/torch_xla/core/dynamo_bridge.py", line 540, in extract_compiled_graph
    extract_internal(fused_module), node.args, None)
  File "/usr/local/lib/python3.8/site-packages/torch_xla-2.2.0+git4e506b7-py3.8-linux-x86_64.egg/torch_xla/core/dynamo_bridge.py", line 338, in extract_internal
    dumb_return_handler, xla_args_need_update) = extract_graph_helper(xla_model)
  File "/usr/local/lib/python3.8/site-packages/torch_xla-2.2.0+git4e506b7-py3.8-linux-x86_64.egg/torch_xla/core/dynamo_bridge.py", line 212, in extract_graph_helper
    assert all(
AssertionError: All tensors should be on xla
```

cc @JackCaoG @wonjoolee95

**Comment by yeounoh** [comment]
> Ok, found another test regression -- `test_mark_sharding_inside_compile` works with `torch` nightly from `11/14/2023` but started failng with the latest (`11/22/2023`):
> 
> ```
> ======================================================================
> FAIL: test_mark_sharding_inside_compile (__main__.DynamoSpmdInferenceTest)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "test/spmd/test_dynamo_spmd.py", line 232, in test_mark_sharding_inside_compile
>     dynamo_res = dynamo_linear(xla_x)
>   File "/usr/local/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
>     return self._call_impl(*args, **kwargs)
>   File "/usr/local/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
>     return forward_call(*args, **kwargs)
>   File "/usr/local/lib/python3.8/site-packages/torch/_dynamo/eval_frame.py", line 489, in _fn
>     return fn(*args, **kwargs)
>   File "/usr/local/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
>     return self._call_impl(*args, **kwargs)
>   File "/usr/local/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
>     return forward_call(*args, **kwargs)
>   File "test/spmd/test_dynamo_spmd.py", line 32, in forward
>     xs.mark_sharding(
>   File "/usr/local/lib/python3.8/site-packages/torch_xla-2.2.0+git4e506b7-py3.8-linux-x86_64.egg/torch_xla/runtime.py", line 78, in wrapper
>     if not using_pjrt():
>   File "/usr/local/lib/python3.8/site-packages/torch_xla-2.2.0+git4e506b7-py3.8-linux-x86_64.egg/torch_xla/runtime.py", line 82, in resume_in_wrapper
>     return fn(*args, **kwargs)
>   File "/usr/local/lib/python3.8/site-packages/torch_xla-2.2.0+git4e506b7-py3.8-linux-x86_64.egg/torch_xla/distributed/spmd/xla_sharding.py", line 499, in mark_sharding
>     num_devices = xr.global_runtime_device_count()
>   File "/usr/local/lib/python3.8/site-packages/torch_xla-2.2.0+git4e506b7-py3.8-linux-x86_64.egg/torch_xla/distributed/spmd/xla_sharding.py", line 499, in resume_in_mark_sharding
>     num_devices = xr.global_runtime_device_count()
>   File "/usr/local/lib/python3.8/site-packages/torch_xla-2.2.0+git4e506b7-py3.8-linux-x86_64.egg/torch_xla/distributed/spmd/xla_sharding.py", line 510, in resume_in_mark_sharding
>     tile_assignment, group_assignment, replication_groups, sharding_type = _extract_op_sharding_specs(
>   File "/usr/local/lib/python3.8/site-packages/torch_xla-2.2.0+git4e506b7-py3.8-linux-x86_64.egg/torch_xla/distributed/spmd/xla_sharding.py", line 413, in _extract_op_sharding_specs
>     def _extract_op_sharding_specs(mesh: Mesh, partition_spec: Tuple):
>   File "/usr/local/lib/python3.8/site-packages/torch/_dynamo/eval_frame.py", line 489, in _fn
>     return fn(*args, **kwargs)
>   File "/usr/local/lib/python3.8/site-packages/torch/_dynamo/external_utils.py", line 17, in inner
>     return fn(*args, **kwargs)
>   File "/usr/local/lib/python3.8/site-packages/torch/_functorch/aot_autograd.py", line 4960, in forward
>     return compiled_fn(full_args)
>   File "/usr/local/lib/python3.8/site-packages/torch/_functorch/aot_autograd.py", line 2017, in g
>     return f(*args)
>   File "/usr/local/lib/python3.8/site-packages/torch/_functorch/aot_autograd.py", line 3164, in runtime_wrapper
>     all_outs = call_func_with_args(
>   File "/usr/local/lib/python3.8/site-packages/torch/_functorch/aot_autograd.py", line 2041, in call_func_with_args
>     out = normalize_as_list(f(args))
>   File "/usr/local/lib/python3.8/site-packages/torch/_functorch/aot_autograd.py", line 2145, in rng_functionalization_wrapper
>     return compiled_fw(args)
>   File "/usr/local/lib/python3.8/site-packages/torch/_functorch/aot_autograd.py", line 2017, in g
>     return f(*args)
>   File "/usr/local/lib/python3.8/site-packages/torch/_dynamo/backends/torchxla.py", line 49, in fwd
>     compiled_graph = bridge.extract_compiled_graph(model, args)
>   File "/usr/local/lib/python3.8/site-packages/torch_xla-2.2.0+git4e506b7-py3.8-linux-x86_64.egg/torch_xla/core/dynamo_bridge.py", line 540, in extract_compiled_graph
>     extract_internal(fused_module), node.args, None)
>   File "/usr/local/lib/python3.8/site-packages/torch_xla-2.2.0+git4e506b7-py3.8-linux-x86_64.egg/torch_xla/core/dynamo_bridge.py", line 338, in extract_internal
>     dumb_return_handler, xla_args_need_update) = extract_graph_helper(xla_model)
>   File "/usr/local/lib/python3.8/site-packages/torch_xla-2.2.0+git4e506b7-py3.8-linux-x86_64.egg/torch_xla/core/dynamo_bridge.py", line 212, in extract_graph_helper
>     assert all(
> AssertionError: All tensors should be on xla
> ```
> 
> cc @JackCaoG @wonjoolee95

This works now.

**Review by wonjoolee95 — APPROVED** [review]
LGTM

**Review by JackCaoG — APPROVED** [review]

**Pull request merged**